Evaluate the argument of polar numbers

normalhuman opened the pull request:
#### Brief description of what is fixed or changed

The argument, like logarithm, is unbranched on the Riemann surface where polar numbers live. So it is computed as unbranched when the argument is a polar number. For example, `arg(exp_polar(4*pi*I))` is returned as `4*pi`, consistent with the purpose of polar numbers: provide continuous evaluation of branched functions such as log and arg.

The current behavior is
```
>>> arg(exp_polar(4*pi*I))
-I*log((re(exp_polar(4*I*pi)) + I*im(exp_polar(4*I*pi))) /
    sqrt(re(exp_polar(4*I*pi))**2 + im(exp_polar(4*I*pi))**2))
```

#### Other comments

Minor changes: `arg` function now uses `as_real_imag` because using re and im just ends up calling the same `as_real_imag` method twice, for real and imaginary parts. 

In the process I also reorganized the import statements, which were under-specific (from sympy import) and scattered throughout the module.

normalhuman committed:
Evaluate the argument of polar numbers

arg(exp_polar(w)) now returns the imaginary part of w, instead of
being mostly un-evaluated. For example, arg(exp_polar(4*pi*I)) is
returned as 4*pi, consistent with the purpose of existence of
polar numbers: provide continuous evaluation of functions such as
log and arg.

The current behavior is
```
>>> arg(exp_polar(4*pi*I))
-I*log((re(exp_polar(4*I*pi)) + I*im(exp_polar(4*I*pi))) /
    sqrt(re(exp_polar(4*I*pi))**2 + im(exp_polar(4*I*pi))**2))
```
which is inconsistent with
```
>>> log(exp_polar(4*pi*I))
4*I*pi
```

For comparison

jksuom commented:
`periodic_argument._get_unbranched` takes `.as_real_imag()[1]`. A comparison might be worth-while.

normalhuman commented:
Thanks for the pointer. This got me thinking more about the complexes.py functions, so I posted an issue #13912.

normalhuman commented:
As an aside, I am yet to find an example where `as_real_imag` returns a different thing from re and im. I understand it's supposed to do a more thorough separation; I just don't see it happening in the examples I tried. 

I also notice that `arg` currently uses `x, y = re(z), im(z)` instead of `x, y = z.as_real_imag()`.

normalhuman commented:
Reading the code of `re` and `im`, I think that their difference from `as_real_imag` is overemphasized. Essentially, `re` splits its argument as a sum and calls `as_real_imag` on each summand. 

So, when both `re` and `im` are needed, calling both instead of using `as_real_imag` is just doubling the work. 

So I think I'm going to defer to unbranched argument implementation for polars, and also change the `arg` to use `_as_real_imag` for ordinary complex numbers.

normalhuman committed:
Refer the computation of arg(polar number) to unbranched_argument

The argument, like logarithm, is unbranched on the Riemann surface
where polar numbers live. So it should be computed as unbranched
when the argument is a polar number.

Also, arg now uses .as_real_imag because using re and im just ends up
calling the same as_real_imag method twice, for real and imaginary
parts.

In the process I also organized the import statements.

normalhuman committed:
Limit arg(polar number) eval to arg(exp_polar)

While it is tempting to send all polar numbers over to
unbranched_argument, the reality is that changing
arg(polar_lift(x)) to periodic_argument(x, oo) is not
really any simplification, and turns out to be
counterproductive for Meijer G integration.

So this commit limits arg evaluation to
arg(exp_polar(w)) = im(w) which is certainly simpler.

jksuom commented:
+1
